### PR TITLE
HLE/Applets: Qt software keyboard implementation and auto result code for SDL

### LINKED ISSUE #1583: Games that still hangs while OS Software Keyboard should appear
**Zoo Resort 3D**

zoo 2

zoo

[LOG](http://pastebin.com/JdtzJ35y)

**Harvest Moon 3D A New Beginning**

reference [video](https://www.youtube.com/watch?v=i5XbZHzOQxc&list=PLMfPU_BxnPBaiCUeaxyf-r-cQG1auezxv)

harvest

[LOG](http://pastebin.com/8XuzqeVb)

### PR DESCRIPTION
**Pull request by valentinvanelslande**
#### Working games & apps:
- [CTRIPCTool](https://gist.github.com/valentinvanelslande/1ccf62f9cef373e99744a6bb3d7b452c)
- [FBI](https://github.com/Steveice10/FBI)
- Ever Oasis
- StreetPass Mii Plaza
- System Settings
- Nintendo Network ID Settings
- Mii Maker
- Miitopia

Screenshot
swkbd

#### TODO
- [ ] Implement multi line, no game that I have has multiline support
- [ ] Implement NOTEMPTY_NOTBLANK and NOTBLANK checks

A lot of code is from #2334
Fixes #1583 #2095 #3382 #3477

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3552)

### CODE REVIEW

File: src/core/settings.h [hunk 1]
```diff
@@ -86,6 +92,13 @@ static const std::array<const char*, NumAnalogs> mapping = {{
 }};
 } // namespace NativeAnalog
 
+struct SwkbdInfo {
+    bool open;
+    std::string text;
+    SwkbdResult return_code;
+    SwkbdConfig config;
+};
```
Comment (wwylele) [inline src/core/settings.h]: This kind of stuff doesn't belong to settings. Settings is the interface for user configurable fields (basically what is stored in the ini file)

File: src/core/hle/applets/swkbd.h [hunk 2]
```diff
@@ -57,6 +155,7 @@ class SoftwareKeyboard final : public Applet {
 
     ResultCode ReceiveParameter(const Service::APT::MessageParameter& parameter) override;
     ResultCode StartImpl(const Service::APT::AppletStartupParameter& parameter) override;
+    void QtImpl();
```
Comment (wwylele) [inline src/core/hle/applets/swkbd.h]: The core shouldn't be aware what frontend it is, so the name `Qt` doesn't make sense here.

File: src/core/settings.h (hunk 1 above)
Comment (valentinvanelslande) [reply src/core/settings.h]: compilation fails if swkbd_info is not there

**Comment by FearlessTobi** [comment]
You are making a change to the frontend so please include a screenshot.

**Comment by valentinvanelslande** [comment]
added

**Comment by wwylele** [comment]
Sorry, I am closing this before it causes unnecessary debate and queueing in the PR list. I don't see this mergeable in the current form and it needs a redesign, mainly in the core<->frontend communication interfaces. Given the poor response you have been giving in your previous PRs, I don't find it useful discussing the issue and fixing them here. So please rework your design and open a new PR when it is ready.

The current design contains the following issues:
 - improper use of Settings module. The applet interface should probably be directly exposed by the `System` class in core.h, or a new interface class dedicated for HLE applet communication
 - unaware of the multi-threading and race conditions issue, and not using proper synchronization method. For example `while(Settings::values.swkbd_info.open) { std::this_thread::sleep_for(std::chrono::nanoseconds{1}); }`

----

I am not against a full HLE applet implementation. It is totally welcome as long as coded properly.

**Comment by jroweboy** [comment]
@valentinvanelslande thanks for the attempt, and I don't mean to be rude when i say this, but instead of telling you to "just code it properly" i'm going to put my other projects on hold temporarily and get the frontend applet interface coded. I plan to build out a basic qt frontend applet, and you are more than welcome to work from there.

Its just every time you try to make a keyboard, you keep submitting the same low quality code that has been submitted before and rejected without addressing any of the feedback or issues with the original code. Additionally, you don't seem to understand what we are saying by making a class for dedicated applet communication, so I figure I'll do that instead.

**Pull request closed without merging**